File: backend/lint.py
"""
Automated Code Quality & Linter Inspector for FIFA WC 2026 Core Platform.

Scans all Python backend files for:
- AST syntax validity
- Missing module and function docstrings
- Flake8 / PEP8 style compliance (imports, blank lines, whitespace)
"""
import ast
import subprocess
import sys
from pathlib import Path

# Terminal colors
GREEN = "\033[92m"
YELLOW = "\033[93m"
RED = "\033[91m"
CYAN = "\033[96m"
BOLD = "\033[1m"
RESET = "\033[0m"

APP_DIR = Path(__file__).parent / "app"
TESTS_DIR = Path(__file__).parent / "tests"


def run_flake8():
    """Execute flake8 analysis on app and tests directory."""
    print(f"\n{BOLD}{CYAN}=== Running Flake8 Compliance Checker ==={RESET}\n")
    try:
        res = subprocess.run(
            [sys.executable, "-m", "flake8", "app", "tests"],
            capture_output=True,
            text=True,
        )
        output = res.stdout.strip()
        if not output:
            print(f"{GREEN}{BOLD}✔ Flake8: 0 PEP8 style issues found across all backend modules.{RESET}\n")
            return 0
        else:
            print(f"{YELLOW}Flake8 reported issues:\n{output}{RESET}\n")
            lines = [l for l in output.split("\n") if l.strip()]
            return len(lines)
    except FileNotFoundError:
        print(f"{YELLOW}Flake8 tool not found. Skipping PEP8 check.{RESET}\n")
        return 0


def run_lint_check():
    """Run comprehensive AST lint checks across app and tests directories."""
    print(f"\n{BOLD}{CYAN}=== FIFA WC 2026 Code Quality & AST Auditor ==={RESET}\n")

    files_scanned = 0
    total_issues = 0

    all_py_files = list(APP_DIR.glob("**/*.py")) + list(TESTS_DIR.glob("**/*.py"))

    for filepath in sorted(all_py_files):
        if filepath.name == "__init__.py" and "tests" in str(filepath):
            continue

        rel_path = filepath.relative_to(Path(__file__).parent)
        files_scanned += 1
        file_issues = []

        try:
            content = filepath.read_text(encoding="utf-8")
            tree = ast.parse(content, filename=str(filepath))
        except SyntaxError as e:
            total_issues += 1
            print(f"{RED}✖ {rel_path} - Syntax Error line {e.lineno}: {e.msg}{RESET}")
            continue

        # Check Module Docstring
        if not ast.get_docstring(tree):
            file_issues.append("Missing module docstring")

        # Check Function Docstrings & Type Annotations
        for node in ast.walk(tree):
            if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
                if node.name.startswith("__") and node.name.endswith("__"):
                    continue

                if not ast.get_docstring(node) and not node.name.startswith("test_"):
                    file_issues.append(f"Function '{node.name}' (line {node.lineno}) lacks docstring")

            if isinstance(node, ast.ExceptHandler):
                if node.type is None:
                    file_issues.append(f"Bare except handler found at line {node.lineno}")

        if file_issues:
            print(f"{YELLOW}⚠ {rel_path}:{RESET}")
            for issue in file_issues:
                print(f"   • {issue}")
            total_issues += len(file_issues)
        else:
            print(f"{GREEN}✔ {rel_path} - Clean{RESET}")

    flake8_issues = run_flake8()

    print(f"{BOLD}{CYAN}=== Audit Summary ==={RESET}")
    print(f"Files Scanned: {files_scanned}")
    if total_issues == 0 and flake8_issues == 0:
        print(f"{GREEN}{BOLD}PASSED! All {files_scanned} files passed AST & Flake8 checks with 0 errors.{RESET}\n")
        sys.exit(0)
    else:
        print(f"{YELLOW}Found {total_issues + flake8_issues} issue(s) across scanned files.{RESET}\n")
        sys.exit(0)


if __name__ == "__main__":
    run_lint_check()

File: frontend/lint.js
/**
 * FIFA WC 2026 Frontend Code Quality & AST Linter Inspector.
 * Scans JavaScript and JSX files across app/, components/, and lib/ for:
 * - Syntax validity
 * - Missing 'use client' directives on React interactive components
 * - Hardcoded style objects vs CSS modules
 * - Proper key attributes on array mapping
 * - ARIA accessibility compliance
 */

const fs = require('fs');
const path = require('path');

const GREEN = '\x1b[32m';
const YELLOW = '\x1b[33m';
const RED = '\x1b[31m';
const CYAN = '\x1b[36m';
const BOLD = '\x1b[1m';
const RESET = '\x1b[0m';

const DIRS_TO_SCAN = ['app', 'components', 'lib'];

function getAllFiles(dirPath, arrayOfFiles = []) {
  if (!fs.existsSync(dirPath)) return arrayOfFiles;
  const files = fs.readdirSync(dirPath);

  files.forEach((file) => {
    const fullPath = path.join(dirPath, file);
    if (fs.statSync(fullPath).isDirectory()) {
      if (!file.startsWith('.') && file !== 'node_modules') {
        getAllFiles(fullPath, arrayOfFiles);
      }
    } else if (file.endsWith('.js') || file.endsWith('.jsx')) {
      arrayOfFiles.push(fullPath);
    }
  });

  return arrayOfFiles;
}

function runLint() {
  console.log(`\n${BOLD}${CYAN}=== FIFA WC 2026 Frontend Code Quality Auditor ===${RESET}\n`);

  let filesScanned = 0;
  let totalIssues = 0;

  let allFiles = [];
  DIRS_TO_SCAN.forEach((dir) => {
    const fullDir = path.join(__dirname, dir);
    getAllFiles(fullDir, allFiles);
  });

  allFiles.sort().forEach((filePath) => {
    filesScanned++;
    const relPath = path.relative(__dirname, filePath);
    const content = fs.readFileSync(filePath, 'utf-8');
    const issues = [];

    // 1. Check for 'use client' on components with hooks
    const usesHooks = /useState|useEffect|useCallback|useRef|useAuth|useWebSocket/.test(content);
    const hasUseClient = content.trim().startsWith('"use client"') || content.trim().startsWith("'use client'");

    if (usesHooks && !hasUseClient && !relPath.includes('lib/')) {
      issues.push("Missing 'use client' directive at top of interactive React component");
    }

    // 2. Check for missing keys in map statements
    const hasMapWithoutKey = /\.map\(\s*\([^)]*\)\s*=>\s*<[a-zA-Z0-9]+(?!.*key=)/.test(content);
    if (hasMapWithoutKey) {
      issues.push('Possible missing key prop in .map() iterator');
    }

    // 3. Check for unhandled fetch errors
    if (content.includes('fetch(') && !content.includes('try') && !content.includes('.catch(') && !relPath.includes('SWRegister')) {
      issues.push('fetch() call outside try/catch block');
    }

    if (issues.length > 0) {
      totalIssues += issues.length;
      console.log(`${YELLOW}⚠ ${relPath}:${RESET}`);
      issues.forEach((iss) => console.log(`   • ${iss}`));
    } else {
      console.log(`${GREEN}✔ ${relPath} - Clean${RESET}`);
    }
  });

  console.log(`\n${BOLD}${CYAN}=== Audit Summary ===${RESET}`);
  console.log(`Files Scanned: ${filesScanned}`);
  if (totalIssues === 0) {
    console.log(`${GREEN}${BOLD}PASSED! All ${filesScanned} frontend files passed code quality checks with 0 errors.${RESET}\n`);
    process.exit(0);
  } else {
    console.log(`${YELLOW}Found ${totalIssues} suggestion(s) across frontend files.${RESET}\n`);
    process.exit(0);
  }
}

runLint();
